### PR TITLE
Cross-Zone Load Balancing for API ELB

### PR DESCRIPTION
**Pull request by austinmoore-**
Adds configuration (`spec.api.loadBalancer.crossZoneLoadBalancing`) to enable cross-zone load balancing on the API ELB. See the added [docs](https://github.com/austinmoore-/kops/blob/96d503e98dfcf36250b995b865d8985fd77ac307/docs/cluster_spec.md#api) for details.

### CODE REVIEW

File: upup/pkg/fi/cloudup/awstasks/loadbalancer_attributes.go [hunk 1]
```diff
@@ -166,6 +166,9 @@ func (_ *LoadBalancer) modifyLoadBalancerAttributes(t *awsup.AWSAPITarget, a, e,
 	if e.ConnectionSettings != nil && e.ConnectionSettings.IdleTimeout != nil {
 		request.LoadBalancerAttributes.ConnectionSettings.IdleTimeout = e.ConnectionSettings.IdleTimeout
 	}
+	if e.CrossZoneLoadBalancing != nil && e.CrossZoneLoadBalancing.Enabled != nil {
```
Comment (granular-ryanbonham) [inline upup/pkg/fi/cloudup/awstasks/loadbalancer_attributes.go]: Should this be an else off the if statement on line 136?

Comment (austinmoore-) [reply upup/pkg/fi/cloudup/awstasks/loadbalancer_attributes.go]: Ah yep 👍

File: pkg/model/awsmodel/api_loadbalancer.go [hunk 2]
```diff
@@ -153,6 +153,13 @@ func (b *APILoadBalancerBuilder) Build(c *fi.ModelBuilderContext) error {
 			Tags: tags,
 		}
 
+		// Only set if specified so we don't change existing logic
+		if lbSpec.CrossZoneLoadBalancing != nil {
```
Comment (KashifSaadat) [inline pkg/model/awsmodel/api_loadbalancer.go]: **Small test:**
- Deploy with `crossZoneLoadBalancing` unset; by default the value on the ELB is set to false
- Set the above in kops spec with value `true`; this is successfully set to true in the ELB Spec
- Revert to original configuration (not explicitly setting in the LoadBalancerSpec to false); no change to the ELB configuration (the value remains true on the ELB resource)

Just a small suggestion that if `crossZoneLoadBalancing` is unset in the LoadBalancerSpec, the ELB property is explicitly set to false (maintains existing behaviour), so the above potential use-case is satisfied.

Comment (austinmoore-) [reply pkg/model/awsmodel/api_loadbalancer.go]: Yeah you're correct. I'll make that change

File: tests/integration/update_cluster/api_elb_cross_zone/in-v1alpha2.yaml [hunk 3]
```diff
@@ -0,0 +1,96 @@
+apiVersion: kops/v1alpha2
```
Comment (justinsb) [inline tests/integration/update_cluster/api_elb_cross_zone/in-v1alpha2.yaml]: It's often easier to throw this into "complex" or another existing test (and it's OK to do so, at least when it's a whole test for a single field)

**Comment by justinsb** [comment]
This looks good - thanks @austinmoore- 

I think at some stage in the future we might well want to make this the default, but we'd have to have a big release note for that so we can hold off for now :-)  The other option is just to make it the default for new clusters that use an ELB.

File: cmd/kops/integration_test.go [hunk 4]
```diff
@@ -80,6 +80,11 @@ func TestComplex(t *testing.T) {
 	runTestAWS(t, "complex.example.com", "complex", "legacy-v1alpha2", false, 1, true, nil)
 }
 
+// TestCrossZone tests that the the cross zone setting on the API ELB is set properly
+func TestCrossZone(t *testing.T) {
+	runTestAWS(t, "crosszone.example.com", "api_elb_cross_zone", "v1alpha2", false, 1, true, nil)
```
Comment (rifelpet) [inline cmd/kops/integration_test.go]: The function signature for `runTestAWS` recently changed in the master branch. You'll want to rebase your branch and add the new boolean argument. Alternatively you can follow justinsb's advice and add the setting to the existing `complex` integration test.

**Comment by austinmoore-** [comment]
@justinsb or @rifelpet can you do another quick review and lgtm for the rebase?

**Comment by rifelpet** [comment]
/lgtm

Thanks  @austinmoore- !

**Comment by rifelpet** [comment]
I'm hoping that #7348 will fix these job failures

**Pull request merged**